### PR TITLE
Add generic schema support

### PR DESCRIPTION
**Pull request by abulte**
This a quick attempt at supporting a generic schema.

The idea is to have a very basic repo (eg https://github.com/abulte/generic-schema) describing the schema, but w/ no data on the schema itself (tableschema file...). This would allow us to reference schemas not supported by our validation mechanism — maybe because they're not based on a technical standard (eg BAL). The schema would have its own page on schema.data.gouv.fr and could be made available on data.gouv.fr too.

I did not go through the hoops of setting up Ruby on my machine so the web part is untested. I'm unsure about what to commit (`aggregateur/(data|cache)`? `web`?) so I kept it to a minimum.

`python main.py` run w/o error.

### CODE REVIEW

File: aggregateur/repertoires.yml [hunk 1]
```diff
@@ -54,3 +54,7 @@ prenoms:
   url: https://github.com/CharlesNepote/liste-prenoms-nouveaux-nes.git
   type: tableschema
   email: charles.nepote@fing.org
+generic-example:
+  url: https://github.com/abulte/generic-schema.git
+  type: generic
+  email: alexandre@example.com
```
Comment (AntoineAugusti) [inline aggregateur/repertoires.yml:60]: Remove this before merging (I know you know)

File: aggregateur/validators.py [hunk 2]
```diff
@@ -342,3 +342,75 @@ def schema_json_data(self):
         with open(self.filepath(self.SCHEMA_FILENAME)) as f:
             self.schema_data = json.load(f)
         return self.schema_data
+
+
+class GenericValidator(BaseValidator):
```
Comment (AntoineAugusti) [inline aggregateur/validators.py:347]: If you can find a better name. Generic vs Base is not very clear. I don't have a good idea, otherwise, leave it this way.

File: aggregateur/validators.py [hunk 3]
```diff
@@ -342,3 +342,75 @@ def schema_json_data(self):
...
+            self.SCHEMA_FILENAME: self.filepath_or_none(self.SCHEMA_FILENAME),
+            "README.md": self.filepath_or_none("README.md"),
+            "SEE_ALSO.md": self.filepath_or_none("SEE_ALSO.md"),
+            "CONTEXT.md": self.filepath_or_none("CONTEXT.md"),
+        }
+
+        if self.is_latest_version():
+            files[self.CHANGELOG_FILENAME] = self.filepath_or_none(
+                self.CHANGELOG_FILENAME
+            )
```
Comment (AntoineAugusti) [inline aggregateur/validators.py:394]: I think you may need to set `has_changelog`to `True` here. It's not pretty, feel feel to refactor other classes if you feel like it.

File: aggregateur/validators.py [hunk 4]
```diff
@@ -342,3 +342,75 @@ def schema_json_data(self):
...
+        keys = ["title", "description", "homepage", "version"]
+        for key in [k for k in keys if k not in self.get_schema_data()]:
+            message = "Key `%s` is a required key and is missing from %s" % (
+                key,
+                self.SCHEMA_FILENAME,
+            )
+            raise exceptions.InvalidSchemaException(self.repo, message)
+
+    def check_schema(self, filename):
+        return True
```
Comment (AntoineAugusti) [inline aggregateur/validators.py:408]: Consider deleting this method.

File: aggregateur/validators.py [hunk 5]
```diff
@@ -342,3 +342,75 @@ def schema_json_data(self):
...
+
+    def validate(self):
+        super().validate()
+        self.check_file_exists(self.SCHEMA_FILENAME)
+        self.check_schema(self.SCHEMA_FILENAME)
+        self.check_extra_keys()
+
+    def extract(self):
+        files = {
+            self.SCHEMA_FILENAME: self.filepath_or_none(self.SCHEMA_FILENAME),
```
Comment (AntoineAugusti) [inline aggregateur/validators.py:385]:
```suggestion
            self.SCHEMA_FILENAME: self.filepath(self.SCHEMA_FILENAME),
```

File: aggregateur/validators.py (hunk 3 above)
Comment (abulte) [reply aggregateur/validators.py:394]: Why? Won't it work just like on TableSchemaValidator (eg use a CHANGELOG.md if there's one)?

Comment (abulte) [reply aggregateur/validators.py]: (changed for the XSD pattern)

File: aggregateur/validators.py (hunk 2 above)
Comment (abulte) [reply aggregateur/validators.py:347]: Generic vs Base is not ultra clear, true, but I think `generic` works well in the rest of the code. I have `meta` or  `dummy` as alternatives, not sure it's better.

File: aggregateur/validators.py (hunk 4 above)
Comment (abulte) [reply aggregateur/validators.py]: Did a little Yaml dance instead.

Comment (abulte) [reply aggregateur/validators.py]: BTW what happens when one of the tags is invalid? The other tags are still published on schema.data.gouv.fr, right?

Comment (AntoineAugusti) [reply aggregateur/validators.py]: Yup, only valid tags are published. If there are errors you'd get an email with details

https://github.com/etalab/schema.data.gouv.fr/blob/56f8207f61d1a153bdbe98c081ab9ad7f1e827c4/aggregateur/main.py#L276-L284

File: aggregateur/validators.py [hunk 6]
```diff
@@ -342,3 +342,86 @@ def schema_json_data(self):
...
+        keys = ["title", "description", "homepage", "version"]
+        for key in [k for k in keys if k not in self.get_schema_data()]:
+            message = "Key `%s` is a required key and is missing from %s" % (
+                key,
+                self.SCHEMA_FILENAME,
+            )
+            raise exceptions.InvalidSchemaException(self.repo, message)
+
+    def open_schema_file(self):
+        data = None
```
Comment (AntoineAugusti) [inline aggregateur/validators.py:408]: My understanding is that you want to return `None` if the YAML file exists but will be empty.

If you return `None`, the `check_extra_keys` will fail because `key in None` raises a `TypeError`. Maybe return `{}`?

File: aggregateur/validators.py [hunk 7]
```diff
@@ -342,3 +342,86 @@ def schema_json_data(self):
...
+        with open(self.filepath(self.SCHEMA_FILENAME)) as f:
+            data = yaml.safe_load(f)
+        return data
+
+    def check_schema(self, filename):
+        try:
+            self.open_schema_file()
+        except yaml.error.YAMLError as e:
```
Comment (AntoineAugusti) [inline aggregateur/validators.py:416]: Is this the only exception that can be raised here? If the file doesn't exist, `open` will raise something else. I think this should never happen because `validate` checks for the existence first

File: aggregateur/validators.py [hunk 8]
```diff
@@ -342,3 +342,86 @@ def schema_json_data(self):
...
+                repr(e),
+            )
+            raise exceptions.InvalidSchemaException(self.repo, message)
+
+    def get_schema_data(self):
+        if self.schema_data is not None:
+            return self.schema_data
+
+        self.schema_data = self.open_schema_file()
+        return self.schema_data
```
Comment (AntoineAugusti) [inline aggregateur/validators.py:427]: I think there is a memoize thing in recent Python versions if you want to go this way

File: aggregateur/validators.py [hunk 9]
```diff
@@ -342,3 +342,86 @@ def schema_json_data(self):
...
+    def check_extra_keys(self):
+        keys = ["title", "description", "homepage", "version"]
+        for key in [k for k in keys if k not in self.get_schema_data()]:
+            message = "Key `%s` is a required key and is missing from %s" % (
+                key,
+                self.SCHEMA_FILENAME,
+            )
+            raise exceptions.InvalidSchemaException(self.repo, message)
+
+    def open_schema_file(self):
```
Comment (AntoineAugusti) [inline aggregateur/validators.py:407]: Is there an advantage to have a method for this? Why did you move it out of `get_schema_data`?

File: web/collections/_documentation/ajouter-un-schema.md [hunk 10]
```diff
@@ -27,6 +27,7 @@ Vous devez ajouter votre dépôt Git en modifiant le fichier [`repertoires.yml`]
     + `tableschema`, pour un schéma au format [Table Schema](https://frictionlessdata.io/specs/table-schema/)
     + `xsd`, pour un schéma au format [XML Schema Definition](https://www.w3.org/TR/xmlschema11-1/) (XSD)
     + `jsonschema`, pour un schéma au format [JSON Schema](https://json-schema.org/)
+    + `generic`, pour un schéma qui ne correspond à aucun autre type supporté
```
Comment (AntoineAugusti) [inline web/collections/_documentation/ajouter-un-schema.md:30]: Maybe explain what it'll contain (documentation + links only)

File: web/collections/_documentation/validation-schemas.md [hunk 11]
```diff
@@ -115,3 +115,18 @@ schemas:
     path: "schemas/bar.json"
     title: "Une description de ce sous-schéma"
 ```
+
+## Validations spécifiques au format générique
```
Comment (AntoineAugusti) [inline web/collections/_documentation/validation-schemas.md:119]: Explain why this format is needed / useful.

Too bad we're introducing this because I think we've got guidance about _why_ a formal schema, and not words only, is useful 😬

File: web/collections/_documentation/validation-schemas.md [hunk 12]
```diff
@@ -115,3 +115,18 @@ schemas:
...
+
+Les dépôts contenant des schémas au format générique subissent les validations supplémentaires suivantes :
+
+- le dépôt doit comporter un fichier `schema.yml` à la racine du dépôt décrivant le schéma ;
+- ce fichier doit être un Yaml valide.
```
Comment (AntoineAugusti) [inline web/collections/_documentation/validation-schemas.md:124]:
```suggestion
- ce fichier doit être un YAML valide.
```
:trollface:

File: aggregateur/validators.py (hunk 6 above)
Comment (AntoineAugusti) [reply aggregateur/validators.py:408]: It may not be the only place where this could cause an issue actually

Comment (abulte) [reply aggregateur/validators.py]: yup I refactored a bit

File: aggregateur/validators.py (hunk 8 above)
Comment (abulte) [reply aggregateur/validators.py]: ✅

File: aggregateur/validators.py (hunk 9 above)
Comment (abulte) [reply aggregateur/validators.py]: not anymore ;-)
